src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,188 @@
+    # "pk" (either zero for columns that are not part of the primary key, or the 1-based index of the column within the primary key)
+    columns = [name for (name, pk) in zip(tableinfo.name, tableinfo.pk) if pk == 0]
+    gpkgbinary = DBInterface.execute(db, "SELECT  $(join(columns, ',')) FROM $tablename;")
+    # length of the GeoPackageBinaryHeader 
+    headerlen = 0
+    geomcollection = map(gpkgbinary) do row
+      # According to https://www.geopackage.org/spec/#r30
+      # A feature table or view SHALL have only one geometry column.
+      columnindex = findfirst(==(Symbol(columnname)), keys(row))
+      rowvals = map(keys(row)[[begin:(columnindex - 1); (columnindex + 1):end]]) do key
+        key, getproperty(row, key)
+      end
+      # get the column of SQL Geometry Binary specified by gpkg_geometry_columns table in column_name field
+      blob = getproperty(row, Symbol(columnname))
+
+      # According to https://www.geopackage.org/spec/#r19
+      # A GeoPackage SHALL store feature table geometries in SQL BLOBs using the Standard GeoPackageBinary format
+      # check the GeoPackageBinaryHeader for the first byte[2] to be 'GP' in ASCII
+      if blob[1:2] != UInt8[0x47, 0x50]
+        @warn "Missing magic 'GP' string in GPkgBinaryGeometry"
+      end
+
+      # create in-memory I/O stream of GeoPackage SQL Geometry Binary Format starting after byte[2] magic = 0x4750;
+      io = IOBuffer(blob)
+      # skip the magic string in header
+      seek(io, 3)
+
+      # bit layout of GeoPackageBinary flags byte
+      # https://www.geopackage.org/spec/#flags_layout
+      # ---------------------------------------
+      # bit # 7 # 6 # 5 # 4 # 3 # 2 # 1 # 0 #
+      # use # R # R # X # Y # E # E # E # B #
+      # ---------------------------------------
+      # R: reserved for future use; set to 0
+      # X: GeoPackageBinary type
+      # Y: empty geometry flag
+      # E: envelope contents indicator code (3-bit unsigned integer)
+      # B: byte order for SRS_ID and envelope values in header
+      flag = read(io, UInt8)
+
+      # 0x07 is a 3-bit mask 0x00001110
+      # left-shift moves the 3-bit mask by one to align with E bits in flag layout
+      # bitwise AND operation isolates the E bits
+      # right-shift moves the E bits by one to align with the least significant bits
+      # results in a 3-bit unsigned integer
+      envelope = (flag & (0x07 << 1)) >> 1
+      
+      # calculate GeoPackageBinaryHeader size in byte stream given extent of envelope:
+      # envelope is [minx, maxx, miny, maxy, minz, maxz], 48 bytes or envelope is [minx, maxx, miny, maxy], 32 bytes or no envelope, 0 bytes
+      # byte[2] magic + byte[1] version + byte[1] flags + byte[4] srs_id + byte[(8*2)×(x,y{,z})] envelope
+      headerlen = iszero(envelope) ? 8 : 8 + 8 * 2 * (envelope + 1)
+
+      # Skip reading the double[] envelope and start reading Well-Known Binary geometry 
+      seek(io, headerlen)
```

Comment:
After you create the io object, I understand that these steps could be encapsulated into a function to skip the header, which consists of the magic number and the envelope.